mathics/builtin/lists.py

```diff
@@ -2012,3 +2012,172 @@ def get_call():
                                              evaluation, same_test))
         result.sort()
         return result
+
+class Fold(Builtin):
+    """
+    <dl>
+    <dt>'Fold[$expr$, $x$, $list$]'
+        <dd>Expression on all elements of list, with initial value of x.
+    <dt>'Fold[$expr$, $list$]'
+        <dd>The same as Fold[$expr$, First[list], Rest[list]]
+    </dl>
+
+    >> Fold[Plus, 5, {1,1,1}]
+     = 8
+    >> Fold[f, 5, {1,2,3}]
+     = f[f[f[5, 1], 2], 3]
+    """
+
+    attributes = ('NHoldRest',)
+
+    rules = {
+        'Fold[exp_, x_, list_List]': 'Module[{},res=x; Do[res = exp[res, list[[i]]], {i, 1, Length[list]}]; res]',
+        'Fold[exp_, list_List]': u'Fold[exp, First[list], Rest[list]]',
+    }
+
+
+class FoldList(Builtin):
+    """
+    <dl>
+    <dt>FoldList[$expr$, $x$, $list$]'
+        <dd>
+        Use expr successive on all elements of list, and return list.
+        </dd>
+        </dt>
+        <dt>FoldList[$expr$, $list$]'
+            <dd>
+            The same as FoldList[$expr$, First[list], Rest[list]].
+            </dd>
+        </dt>
+    </dl>
+    >> FoldList[f, x, {1,2,3}]
+     = {x, f[x, 1], f[f[x, 1], 2], f[f[f[x, 1], 2], 3]}
+    >> FoldList[Times, {1,2,3}]
+     = {1, 2, 6}
+    """
+
+    rules = {
+        'FoldList[exp_, x_, list_List]': 'Prepend[Table[Fold[exp, x, Take[list, i]], {i, 1, Length[list]}], x]',
+        'FoldList[exp_, list_List]': 'FoldList[exp, First[list], Rest[list]]',
+    }
+
+
+class Accumulate(Builtin):
+    """
+    <dl>
+    <dt>Accumulate[$list$]'
+        <dd>
+        Accumulate values from list and return new list
+        </dd>
+        </dt>
+    </dl>
+    >> Accumulate[{1,2,3}]
+     = {1, 3, 6}
+    """
+
+    rules = {
+        'Accumulate[list_List]': 'FoldList[Plus, list]'
+    }
+
+
+class Total(Builtin):
+    """
+    <dl>
+    <dt>Total[$list$]'
+        <dd>
+        Add all values up to calculate total
+        Equivalent to Fold[Plus, $list$] or Apply[Plus, $list$]
+        </dd>
+        </dt>
+    </dl>
+    >> Total[{1,2,3}]
+     = 6
+    """
```

Comment:
Is this where we need the proper `Flatten` functionality; for the other `Total` form?